### PR TITLE
support use memory buffer as input to read multi-page image

### PR DESCRIPTION
**Pull request by Harvey-Huang**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by Harvey-Huang** [comment]
The imreadmulti method only supports std::string as an input, and the file path(std::string) is not support Unicode, imread is not support Unicode file path too, but I can use imdecode which allow to decode image using memory buffer as input.

**Comment by asmorkalov** [comment]
Thanks a lot for contribution. Please add tests for the new API. Please take a look on existing tests for `imreadmulti` as reference.

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/imgcodecs/src/loadsave.cpp [hunk 1]
```diff
@@ -967,6 +967,134 @@ Mat imdecode( InputArray _buf, int flags, Mat* dst )
...
+        return 0;
+
+    if (count < 0) {
+        count = std::numeric_limits<int>::max();
+    }
+
+    if (!decoder->setSource(buf_row))
+    {
+        filename = tempfile();
+        FILE* f = fopen(filename.c_str(), "wb");
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:991]: where is this file removed?

Comment (asmorkalov) [reply modules/imgcodecs/src/loadsave.cpp:994]: @alalek I added file removal and sqaushed commits. The PR is ready to be merged.

**Comment by Harvey-Huang** [comment]
Hi Asmorkalov and Alalek, thanks for the comments and code changed, I just saw this item update today, I forgot to remove the temporary file.

**Comment by vpisarev** [comment]
I have no showstopper complains about this PR, I just wonder if we can rework this patch to use the newly introduced ImageCollection API?

**Comment by asmorkalov** [comment]
New API with iterators can be extended to support in memory processing, but not re-used. I propose to merge the patch to have consistent API ans then extend iterators approach.

**Review by vpisarev — APPROVED** [review]

**Pull request merged**